File: 算法/js/排序/heapSort.js
var heapify = function (array, heapSize, i) {
    var left = i * 2 + 1,
    right = i * 2 + 2,
    largest = i;
    if(left < heapSize && array[left] > array[largest]){
        largest = left;
    }
    if(right < heapSize && array[right] > array[largest]){
        largest = right;
    }
    if(largest !== i){
        [array[i],array[largest]] = [array[largest],array[i]];
        heapify(array,heapSize,largest);
    }
    return array;
};

var buildHeap = function(array){
    var heapSize = array.length;
    for(var i=Math.floor(array.length / 2);i>=0;i--){
        heapify(array,heapSize,i);
    }
    return array;
};

function heapSort(array){
    var heapSize = array.length;
    buildHeap(array);

    while(heapSize > 1){
        heapSize--;
        [array[0],array[heapSize]] = [array[heapSize],array[0]];
        heapify(array,heapSize,0);
    }
    return array;
}

console.log(heapSort([3, 5, 1, 6, 4, 7, 2]))

File: 算法/python/排序/heapSort.py
def heapify(arr,i,heapSize):
    left = 2 * i + 1
    right = 2 * i + 2
    largest = i

    if left < heapSize and arr[largest] < arr[left]:
        largest = left
    if right < heapSize and arr[largest] < arr[right]:
        largest = right
    if largest != i:
        arr[i],arr[largest] = arr[largest],arr[i]
        heapify(arr,largest,heapSize)

def buildHeap(arr):
    heapSize = len(arr)
    for i in range(len(arr)//2,-1,-1):
        heapify(arr,i,heapSize)

def heapSort(arr):
    heapSize = len(arr)
    buildHeap(arr)

    while heapSize > 1:
        heapSize -= 1
        arr[0],arr[heapSize] = arr[heapSize],arr[0]
        heapify(arr,0,heapSize)
    return arr


print(heapSort([1,5,6,8,2,9,10,15,1,19]))